CoreScriptsRoot/ServerCoreScripts/ClientChat/ChatMain.lua

```diff
@@ -0,0 +1,735 @@
+MessageLabelCreator:RegisterSpeakerDatabase(SpeakerDatabase)
+
+
+
+
+
+
+
+
+
+
+--////////////////////////////////////////////////////////////////////////////////////////////
+--////////////////////////////////////////////////////////////// Code to do chat window fading
+--////////////////////////////////////////////////////////////////////////////////////////////
+local function CheckIfPointIsInSquare(checkPos, topLeft, bottomRight)
+	return (topLeft.X <= checkPos.X and checkPos.X <= bottomRight.X and 
+		topLeft.Y <= checkPos.Y and checkPos.Y <= bottomRight.Y)
+end
+
+local backgroundIsFaded = false
+local textIsFaded = false
+local lastFadeTime = 0
+local backgroundFadeTimer = 0
+local textFadeTimer = 30
+
+local fadedChanged = Instance.new("BindableEvent")
+local mouseStateChanged = Instance.new("BindableEvent")
+local chatBarFocusChanged = Instance.new("BindableEvent")
+
+local defaultFadingTime = 0.3
+
+local function DoBackgroundFadeIn(setFadingTime)
+	lastFadeTime = tick()
+	backgroundIsFaded = false
+	fadedChanged:Fire()
+	ChatWindow:FadeInBackground((setFadingTime or defaultFadingTime))
+
+	local currentChannelObject = ChatWindow:GetCurrentChannel()
+	if (currentChannelObject) then
+		local Scroller = currentChannelObject.Scroller
+		Scroller.ScrollingEnabled = true
+		Scroller.ScrollBarThickness = moduleChatChannel.ScrollBarThickness
+	end
+end
+
+local function DoBackgroundFadeOut(setFadingTime)
+	lastFadeTime = tick()
+	backgroundIsFaded = true
+	fadedChanged:Fire()
+	ChatWindow:FadeOutBackground((setFadingTime or defaultFadingTime))
+
+	local currentChannelObject = ChatWindow:GetCurrentChannel()
+	if (currentChannelObject) then
+		local Scroller = currentChannelObject.Scroller
+		scrollBarThickness = Scroller.ScrollBarThickness
+		Scroller.ScrollingEnabled = false
+		Scroller.ScrollBarThickness = 0
+	end
+end
+
+local function DoTextFadeIn(setFadingTime)
+	lastFadeTime = tick()
+	textIsFaded = false
+	fadedChanged:Fire()
+	ChatWindow:FadeInText((setFadingTime or defaultFadingTime) * 0)
+end
+
+local function DoTextFadeOut(setFadingTime)
+	lastFadeTime = tick()
+	textIsFaded = true
+	fadedChanged:Fire()
+	ChatWindow:FadeOutText((setFadingTime or defaultFadingTime))
+end
+
+local function DoFadeInFromNewInformation()
+	DoTextFadeIn()
+	DoBackgroundFadeIn()
+end
+
+local function InstantFadeIn()
+	DoBackgroundFadeIn(0)
+	DoTextFadeIn(0)
+end
+
+local function InstantFadeOut()
+	DoBackgroundFadeOut(0)
+	DoTextFadeOut(0)
+end
+
+local function DealWithCoreGuiEnabledChanged(enabled)
+	if (moduleApiTable.Visible) then
+		if (enabled) then
+			InstantFadeIn()
+		else
+			InstantFadeOut()
+		end
+	end
+end
+
+local mouseIsInWindow = nil
+local function UpdateFadingForMouseState(mouseState)
+	mouseIsInWindow = mouseState
+
+	mouseStateChanged:Fire()
+
+	if (ChatBar:IsFocused()) then return end
+
+	if (mouseState) then
+		DoBackgroundFadeIn()
+		DoTextFadeIn()
+	else
+		DoBackgroundFadeIn()
+	end
+end
+
+
+local last = 0
+spawn(function()
```

Comment:
The preferred way to do RenderStepped code like this is rather to call `BindToRenderStep` on the RunService